### PR TITLE
Making CaloRecHit cuda friendly

### PR DESCRIPTION
**Pull request by vkhristenko**
modified CaloRecHit in order to be able to use on the device.
Prerequisite to use Hcal RecHits

- move impl to header
- add constexpr
- add tests

@fwyzard @makortel

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: DataFormats/CaloRecHit/interface/CaloRecHit.h [hunk 1]
```diff
@@ -4,27 +4,63 @@
 #include "DataFormats/DetId/interface/DetId.h"
 #include <ostream>
 
+#ifdef __CUDA_ARCH__
+  __constant__
+#else
+  constexpr
+#endif
+uint32_t masks[] = {
```
Comment (slava77) [inline DataFormats/CaloRecHit/interface/CaloRecHit.h]: this is named to simply to be a global symbol.
Please move inside CaloRecHit  or give this a more creative name.

Comment (vkhristenko) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: how about anonymous namespace?

File: DataFormats/CaloRecHit/interface/CaloRecHit.h [hunk 2]
```diff
@@ -4,27 +4,63 @@
...
-  uint32_t aux() const { return aux_; }
+  constexpr float energy() const { return energy_; }
+  constexpr void setEnergy(float energy) { energy_=energy; }
+  constexpr float time() const { return time_; }
+  constexpr void setTime(float time) { time_=time; }
+  constexpr const DetId& detid() const { return id_; }
+  constexpr uint32_t flags() const { return flags_; }
+  constexpr void setFlags(uint32_t flags) { flags_=flags; }
+  constexpr void setFlagField(uint32_t value, int base, int width=1) {
+#ifdef __CUDA_ARCH__
```
Comment (fwyzard) [inline DataFormats/CaloRecHit/interface/CaloRecHit.h]: are you sure we need these `#ifdef  __CUDA_ARCH__` blocks ? 
`std::min` and `std::max` should be `constexpr`, so they should be available in device code as well.

File: DataFormats/CaloRecHit/interface/CaloRecHit.h [hunk 3]
```diff
@@ -4,27 +4,65 @@
 #include "DataFormats/DetId/interface/DetId.h"
 #include <ostream>
 
+namespace {
```
Comment (slava77) [inline DataFormats/CaloRecHit/interface/CaloRecHit.h]: this should not be used in header files

Comment (vkhristenko) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: ok, did not know... thx

i renamed, as __constant__ (memory specifier) can not be marked on a class member

File: DataFormats/CaloRecHit/interface/CaloRecHit.h (hunk 2 above)
Comment (vkhristenko) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: true, removed

File: DataFormats/CaloRecHit/interface/CaloRecHit.h [hunk 4]
```diff
@@ -4,27 +4,52 @@
...
 /** \class CaloRecHit
  * 
  *\author J. Mans - Minnesota
  */
 class CaloRecHit {
 public:
-  CaloRecHit(); // for persistence
-  explicit CaloRecHit(const DetId& id, float energy, float time, uint32_t flags = 0, uint32_t aux=0);
+  constexpr CaloRecHit() : energy_(0), time_(0), flags_(0), aux_(0) {}
+  constexpr explicit CaloRecHit(const DetId& id, float energy, float time,
```
Comment (makortel) [inline DataFormats/CaloRecHit/interface/CaloRecHit.h]: While at it, maybe drop the `explicit` as it has no effect here?

File: DataFormats/CaloRecHit/test/test_calo_rechit.cu [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+int main(int argc, char** argv) {
+    int nDevices;
+    cudaGetDeviceCount(&nDevices);
+    std::cout << "nDevices = " << nDevices << std::endl;
+
+    if (nDevices > 0)
+        test_calo_rechit();
+
+    std::cout << "all good!" << std::endl;
+}
```
Comment (makortel) [inline DataFormats/CaloRecHit/test/test_calo_rechit.cu]: Could you add `return 0`?

**Review by makortel — COMMENTED** [review]
Two minor comments.

File: DataFormats/CaloRecHit/interface/CaloRecHit.h (hunk 4 above)
Comment (vkhristenko) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: done

File: DataFormats/CaloRecHit/test/test_calo_rechit.cu (hunk 5 above)
Comment (vkhristenko) [reply DataFormats/CaloRecHit/test/test_calo_rechit.cu]: done

File: DataFormats/CaloRecHit/interface/CaloRecHit.h (hunk 4 above)
Comment (fwyzard) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: > While at it, maybe drop the `explicit` as it has no effect here?

Actually it has an effect, but I don't think we ever use that.
With `explicit` you can construct a `CaloRecHit` with any of these syntaxes:
```c++
CaloRecHit crh1{ id, energy, time };
CaloRecHit crh2( id, energy, time );
CaloRecHit crh3 = CaloRecHit{ id, energy, time };
CaloRecHit crh4 = CaloRecHit( id, energy, time );
```
(although these last two are better written with `auto`:
```c++
auto crh3 = CaloRecHit{ id, energy, time };
auto crh4 = CaloRecHit( id, energy, time );
```
).

Without `explicit` you could use also this:
```c++
CaloRecHit crh5 = { id, energy, time };
```

As I said, not something we are likely to use anyway.

Comment (fwyzard) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: See more details on [Stakoverflow](https://stackoverflow.com/a/39122237/2050986).

Comment (makortel) [reply DataFormats/CaloRecHit/interface/CaloRecHit.h]: Thanks @fwyzard, I stand corrected (I had missed this detail of C++11). Then I don't have an opinion whether `explicit` should be kept here or not.

**Comment by slava77** [comment]
this looks superseded by #23495.
Can it be closed?

**Comment by vkhristenko** [comment]
@slava77 i thought since this part is isolated completely, it's better to have a separate pr... and to get it merged faster, but in principle https://github.com/cms-sw/cmssw/pull/23495 has branch of this pr merged

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@vkhristenko 
I'd still prefer to have this closed
this way I can actually start a review of  #23495

**Comment by slava77** [comment]
-1

superseded by #23495

**Comment by fabiocos** [comment]
@vkhristenko @slava77 if this PR is superseded by a newer one, should it be just closed?

**Comment by vkhristenko** [comment]
@fabiocos , yes, closing

**Pull request closed without merging**